plugin-server/src/worker/ingestion/person-state.ts

```diff
@@ -231,30 +227,41 @@ export class PersonState {
         }
     }
 
-    private updatedPersonProperties(personProperties: Properties): Properties {
-        const updatedProperties = { ...personProperties }
+    private updatedPersonProperties(person: Partial<Person>): Partial<Person> {
+        const updatedPerson = {
+            properties: person.properties || {},
+            properties_last_updated_at: person.properties_last_updated_at || {},
+            properties_last_operation: person.properties_last_operation || {},
+        }
 
         const properties: Properties = this.eventProperties['$set'] || {}
         const propertiesOnce: Properties = this.eventProperties['$set_once'] || {}
         const unsetProperties: Array<string> = this.eventProperties['$unset'] || []
+        const timestampIso = this.timestamp.toISO()
 
         // Figure out which properties we are actually setting
         Object.entries(propertiesOnce).map(([key, value]) => {
-            if (typeof personProperties[key] === 'undefined') {
-                updatedProperties[key] = value
+            if (!person.properties || typeof person.properties[key] === 'undefined') {
+                updatedPerson.properties[key] = value
+                updatedPerson.properties_last_operation[key] = PropertyUpdateOperation.SetOnce
+                updatedPerson.properties_last_updated_at[key] = timestampIso
             }
         })
         Object.entries(properties).map(([key, value]) => {
-            if (personProperties[key] !== value) {
-                updatedProperties[key] = value
+            if (!person.properties || person.properties[key] !== value) {
+                updatedPerson.properties[key] = value
+                updatedPerson.properties_last_operation[key] = PropertyUpdateOperation.Set
+                updatedPerson.properties_last_updated_at[key] = timestampIso
             }
         })
 
         unsetProperties.forEach((propertyKey) => {
-            delete updatedProperties[propertyKey]
+            delete updatedPerson.properties[propertyKey]
+            updatedPerson.properties_last_operation[propertyKey] = PropertyUpdateOperation.Unset
```

Comment:
Please write proper unit tests for all new plugin server code. This is untested.